src/stim/util_top/reference_sample_tree.cc

```diff
@@ -0,0 +1,179 @@
+#include "stim/util_top/reference_sample_tree.h"
+
+using namespace stim;
+
+bool ReferenceSampleTree::empty() const {
+    if (repetitions == 0) {
+        return true;
+    }
+    if (!prefix_bits.empty()) {
+        return false;
+    }
+    for (const auto &child: suffix_children) {
+        if (!child.empty()) {
+            return false;
+        }
+    }
+    return true;
+}
+
+void ReferenceSampleTree::flatten_and_simplify_into(std::vector<ReferenceSampleTree> &out) const {
+    if (repetitions == 0) {
+        return;
+    }
+
+    std::vector<ReferenceSampleTree> flattened;
+    if (!prefix_bits.empty()) {
+        flattened.push_back(ReferenceSampleTree{
+            .prefix_bits = prefix_bits,
+            .suffix_children = {},
+            .repetitions = 1,
+        });
+    }
+    for (const auto &child : suffix_children) {
+        child.flatten_and_simplify_into(flattened);
+    }
+
+    // Fuse children.
+    auto &f = flattened;
+    for (size_t k = 0; k < f.size(); k++) {
+        // Combine children with identical contents by adding their rep counts.
+        while (k + 1 < f.size() && f[k].prefix_bits == f[k + 1].prefix_bits && f[k].suffix_children == f[k + 1].suffix_children) {
+            f[k].repetitions += f[k + 1].repetitions;
+            f.erase(f.begin() + k + 1);
+        }
+
+        // Fuse children with unrepeated contents if they can be fused.
+        while (k + 1 < f.size() && f[k].repetitions == 1 && f[k].suffix_children.empty() && f[k + 1].repetitions == 1) {
+            f[k].suffix_children = std::move(f[k + 1].suffix_children);
+            f[k].prefix_bits.insert(f[k].prefix_bits.end(), f[k + 1].prefix_bits.begin(), f[k + 1].prefix_bits.end());
+            f.erase(f.begin() + k + 1);
+        }
+    }
+
+    if (repetitions == 1) {
+        // Un-nest all the children.
+        for (auto &e : flattened) {
+            out.push_back(e);
+        }
+    } else if (flattened.size() == 1) {
+        // Merge with single child.
+        flattened[0].repetitions *= repetitions;
+        out.push_back(std::move(flattened[0]));
+    } else if (flattened.empty()) {
+        // Nothing to report.
+    } else if (flattened[0].suffix_children.empty() && flattened[0].repetitions == 1) {
+        // Take payload from first child.
+        auto result = std::move(flattened[0]);
+        flattened.erase(flattened.begin());
```

Comment:
What do you mean? Doesn't `erase` move the elements to fill in the hole, instead of copying them? There's supposed to be no copying here.

Reply:
Whether its deemed a "copy" or a "move" (constructor/assignment semantic labels), its still O(n) work where data is copied. Move is just a pattern for a shallow copy + invalidation of the source.

Reply:
The `erase()` call can be removed if you did
```
        result.suffix_children.insert(
            result.suffix_children.end(),
            std::make_move_iterator(fused.begin() + 1),
            std::make_move_iterator(fused.end()));
```
instead of `result.suffix_children = std::move(fused);`.